File: game-server/src/handlers/battle/start_cocoon_stage.rs
use crate::utils::time;
use configs::srtools::SrToolsConfig;
use sr_proto::{SceneBattleInfo, StartCocoonStageCsReq, StartCocoonStageScRsp};
use sr_proto::prost::Message;

pub async fn handle(req: &[u8]) -> StartCocoonStageScRsp {
    let cfg = SrToolsConfig::from_file("_configs_/config.json").await;
    let req = StartCocoonStageCsReq::decode(req).unwrap_or_default();

    let battle_info = SceneBattleInfo {
        buff_list: cfg.get_battle_buffs(),
        battle_id: cfg.get_battle_id(),
        stage_id: cfg.get_stage_id(),
        rounds_limit: cfg.get_cycle_count(),
        monster_wave_list: cfg.get_battle_waves(),
        battle_avatar_list: cfg.get_battle_avatars(),
        logic_random_seed: time::cur_timestamp_for_seed(),
        ..Default::default()
    };

    StartCocoonStageScRsp {
        retcode: 0,
        cocoon_id: req.cocoon_id,
        prop_entity_id: req.prop_entity_id,
        wave: 1,
        battle_info: Some(battle_info),
    }
}

File: gameserver/handler/battle/StartCocoonStage.go
package battle

import (
	"fmt"
	"time"

	"github.com/yuvlian/qingque-sr/config"
	"github.com/yuvlian/qingque-sr/config/srtools"
	"github.com/yuvlian/qingque-sr/gameserver/session"
	"github.com/yuvlian/qingque-sr/pb"
	"github.com/yuvlian/qingque-sr/pb/cid"
)

func StartCocoonStage(s *session.Session) error {
	var req pb.StartCocoonStageCsReq
	if err := s.DecodeReq(&req); err != nil {
		return err
	}

	wasReloaded, err := srtools.Reload(config.SRToolsLiteFilePath)
	if err != nil {
		fmt.Println("failed to reload SRToolsLite:", err)
	}
	if wasReloaded {
		fmt.Println("reloaded SRToolsLite!")
	} else {
		fmt.Println("using already loaded SRToolsLite.")
	}

	battleAvatars, errList := srtools.Loaded.GetBattleAvatars()
	if len(errList) > 0 {
		for _, err := range errList {
			fmt.Println(err)
		}
	}

	battleInfo := &pb.SceneBattleInfo{
		BuffList:         srtools.Loaded.GetBattleBuffs(),
		BattleId:         srtools.Loaded.GetBattleID(),
		StageId:          srtools.Loaded.GetStageID(),
		RoundsLimit:      srtools.Loaded.GetCycleCount(),
		MonsterWaveList:  srtools.Loaded.GetBattleWaves(),
		BattleAvatarList: battleAvatars,
		LogicRandomSeed:  uint32(time.Now().Unix() / 1000),
	}

	rsp := &pb.StartCocoonStageScRsp{
		CocoonId:     req.CocoonId,
		PropEntityId: req.PropEntityId,
		Wave:         1,
		BattleInfo:   battleInfo,
	}
	return s.Send(cid.StartCocoonStageScRsp, rsp)
}
